pkg/cluster/config.go

```diff
@@ -57,55 +57,71 @@ type Config struct {
 	Tiller bool
 }
 
+// iterate func map for config template
+func iterate(start, end int) (stream chan int) {
+	stream = make(chan int)
+	go func() {
+		for i := start; i <= end; i++ {
+			stream <- i
+		}
+		close(stream)
+	}()
+	return
+}
+
 // GenerateKindConfig creates kind config file and returns its path
-func GenerateKindConfig(cl *Config, configDir string, box *packr.Box) (string, error) {
-	kindConfigFileTemplate, err := box.Resolve("tpl/cluster-config.yaml")
+func GenerateKindConfig(config *Config, configDir string, box *packr.Box) (string, error) {
+	templateFile := "tpl/cluster-config.yaml"
+	kindConfigTemplateFile, err := box.Resolve(templateFile)
 	if err != nil {
-		return "", err
+		return "", errors.Wrapf(err, "failed to find kind config template file %q", templateFile)
 	}
 
-	t, err := template.New("config").Funcs(template.FuncMap{"iterate": iterate}).Parse(kindConfigFileTemplate.String())
+	kindConfigTemplate, err := template.New("config").Funcs(template.FuncMap{"iterate": iterate}).Parse(kindConfigTemplateFile.String())
 	if err != nil {
-		return "", err
+		return "", errors.Wrapf(err, "failed to parse kind config template file %q", templateFile)
 	}
 
-	kindConfigFilePath := filepath.Join(configDir, "kind-config-"+cl.Name+".yaml")
-	f, err := os.Create(kindConfigFilePath)
+	kindConfigFilePath := filepath.Join(configDir, "kind-config-"+config.Name+".yaml")
+	kindConfigFile, err := os.Create(kindConfigFilePath)
 	if err != nil {
-		return "", err
+		return "", errors.Wrapf(err, "failed to create kind config file %q", kindConfigFilePath)
 	}
 
-	err = t.Execute(f, cl)
+	err = kindConfigTemplate.Execute(kindConfigFile, config)
 	if err != nil {
-		return "", err
+		os.Remove(kindConfigFilePath)
+		return "", errors.Wrapf(err, "failed to generated kind config file %q", kindConfigFilePath)
 	}
 
-	if err := f.Close(); err != nil {
-		return "", err
+	if err := kindConfigFile.Close(); err != nil {
+		os.Remove(kindConfigFilePath)
```

Comment:
> Not sure it makes sense to delete the file, from the docs it seems that close will return an error if the file is already closed in which case I'm not sure you'd want to delete it.

I'm not sure I follow. We just successfully created the file on L86 so it wouldn't be already closed when we hit L97. 

> 
> Also, shouldn't we have the close with defer right after the file is opened (at create)?

We could but then we wouldn't be able to return an error if `Close` fails.

Reply:
> 
> > Also, shouldn't we have the close with defer right after the file is opened (at create)?
> 
> We could but then we wouldn't be able to return an error if `Close` fails.

Right, that a good point. We could do it in a way that also catches that error if it happens [1].
Anyway this is not a major point just a thought on how should we be making sure the code is more resilient by using go's defer mechanism. This way you wouldn't need to remove the file manually on L93 since it would be handled by the defer.

[1] https://blog.learngoprogramming.com/5-gotchas-of-defer-in-go-golang-part-iii-36a1ab3d6ef1

Reply:
OK I changed it to use defer and named params. TBH, I'm not really crazy about it but I can live with it.